Add end to end test devoted to ctest

hippo91 opened the pull request:
This adds an end-to-end test suite for ctest

This PR adds an end-to-end test suite dedicated to the ctest part of the extension, under test/end-to-end-tests/single-root-ctest/.

### What's included:

- A new single-root-ctest end-to-end test project that exercises cmake.ctest against three scenarios (2 successes, 2 successes + 1 failure, 3 failures) across 4 modes each (with/without parallel jobs, with/without test-suite delimiter).
- Hidden LinuxBase / WindowsBase configure presets that set the compilers per OS, mixed with hidden test scenario presets.
- A endToEndTestsSingleRootCTest yarn script and a matching Run single-root-ctest Tests launch configuration.
- The new suite is wired into all three PR CI workflows (ci-main-linux.yml, ci-main-mac.yml, ci-main.win.yml).
- Purpose
- When making changes to the ctest part of the extension, there was no end-to-end coverage to fall back on. This PR fills that gap so future ctest changes can be made with more confidence.

## Note
Earlier revisions of this PR also modified src/ctest.ts and src/cmakeProject.ts; both source-file changes were removed during review. The current diff is test-only.

Commit:
Adds a new end to end test directory devoted to ctest tests

Commit:
Fix non stopping breakpoints

Commit:
Adds scripts entry for new end to end ctest

Commit:
Adds configurations for new end to end ctest test

Commit:
Do not forbid test refreshing even in test mode

hippo91 committed:
Fix end to end test devoted to ctest

hippo91 committed:
Test is functionnal when running through debug

Commit:
Add "allowParallelJobs" in the config

Commit:
Removes useless files

Commit:
Replaces TestCleanup target with simpler call in the teardown method

Commit:
Add cmake.ctest.allowParallelJobs setting

Commit:
Upgrade cmake minimum required version to avoid warnings

Commit:
Sets up two tests. One with parallel jobs enabled and the other without

Commit:
Remove useless file

Commit:
Adds two tests dealing with testSuiteDelimiter

hippo91 committed:
Remove generated test file between each test

hippo91 committed:
Do not delete file if it is not present

Commit:
Fix test suite delimiter

Commit:
Do not add a comma if no text is appended

Commit:
Delete test file results before the test

Commit:
Create cleanUpTestResultFiles to factorize code

Commit:
Adds explicit message to failing assertions

Commit:
Fix comment

Commit:
Generalize output writing

hippo91 committed:
Protect empy successor when dealing with last element of the container

hippo91 committed:
Adds doc and format

hippo91 committed:
Move suite initialization in the right method

hippo91 committed:
Await for settings change promise

hippo91 committed:
Set CMT_TESTING to one

hippo91 committed:
Restore previous state

hippo91 committed:
Renames files and targets

hippo91 committed:
Create a TestUtils lib to factorize code

hippo91 committed:
Start templating test file

Commit:
Move cmake stuff in its own module

hippo91 committed:
Adds doc

hippo91 committed:
Set tests definition in CMake presets

hippo91 committed:
Renames preset

Commit:
Remove generated test from git

Commit:
Fix generate_test_source_file function and return value

Commit:
Introduces use of common directory to store test result

Commit:
Defines common tests directory in cmake cache variables

Commit:
Add and fix doc. Renames variable.

Commit:
First ry to retrieve CMakePresets name in the test

hippo91 committed:
Remove tmp test dir, its location being read from CMakePresets.

hippo91 committed:
Reword test labels

hippo91 committed:
Format

hippo91 committed:
Turns method into free function to be reused in other test suites

hippo91 committed:
Removes unused import

hippo91 committed:
Adds another test suite to check test failure detection

hippo91 committed:
Typo

hippo91 committed:
Adds comments

hippo91 committed:
Factorize code

hippo91 committed:
Add newline at eof

hippo91 committed:
Typos

hippo91 committed:
Adds suite test to check when all tests fail

hippo91 committed:
Adds doc. Removes comment

hippo91 committed:
Moves test env initialization into commonSetup method to avoid code duplication

hippo91 committed:
Adds doc

hippo91 committed:
Remove dead code

gcampbell-msft committed:
Merge branch 'main' into add_end_to_end_ctest_tests

gcampbell-msft reviewed (changes requested):
I haven't reviewed the test project itself yet, I hope to get to it later today or next week, but I have some feedback now that could be addressed. 

Also, it is **essential** that this test gets added to the PR tests, like here: https://github.com/microsoft/vscode-cmake-tools/blob/5107f7edfb4a748c0ef1606bf4db2a2f54cb491e/.github/workflows/ci-main.win.yml#L51C4-L58C28

gcampbell-msft commented on test/end-to-end-tests/single-root-ctest/project-folder/CMakePresets.json:
```diff
@@ -0,0 +1,44 @@
+      "description": "Sets generator, build and install directory, vcpkg",
+      "generator": "Ninja",
+      "binaryDir": "${workspaceFolder}/build",
+      "cacheVariables": {
```
We need to ensure that we define the CMAKE_C_COMPILER and CMAKE_CXX_COMPILER, depending on what OS this is run on.

hippo91 replied:
@gcampbell-msft i added hidden configure preset that hold compiler specifications for linux and windows systems. These presets are then mixed up with those setting up the tests to have all cases covered.

gcampbell-msft commented:
@hippo91 I apologize for taking so long to start reviewing these, we appreciate your patience. 

We GREATLY appreciate your contribution to our engineering system, so we'd love to work with you and iterate on this PR to make these testing contributions, 

A couple of things: 
1. These tests should be integrated into our PR build pipelines so that we can test them in each PR
2. To do that, they should be set up to be cross-platform so that they can be run on each platform, since ctest is available on all three OS's. This will require modifications to the presets to define things like the CMAKE_C_COMPILER, CMAKE_CXX_COMPILER, and possibly using different generators on each OS. 
3. Currently, are you able to get these to work locally on Windows? I notice that I hit errors when trying to open and run in Windows. 

Again, we greatly appreciate this contribution and hope to iterate on the PR with you so that we can improve our engineering system. 

Thanks! Let us know if you have any questions.

snehara99 committed:
Updated Issue Template Selection (#4335)

* updated templates

* wording updates

* updated changelog

* template updates

hippo91 committed:
Adds endToEndTestsSingleRootCTest to the list of tests to run in the CI pipelines for windows, linux, and macos.

hippo91 committed:
Remove launch configuration modification that was not made on purpose

hippo91 commented on .vscode/launch.json:
```diff
@@ -105,6 +105,7 @@
             ],
             "sourceMaps": true,
             "outFiles": [
+                "${workspaceFolder}/dist/**/*.js",
```
@gcampbell-msft you are right it was an error.

hippo91 committed:
Fix 3232 (#4223)

* Adds the getAllCMakeProjects method

* Loop over all CMake projects of the workspace to provide tasks.

Also pass the workspaceFolder to the provideTask method

* Add workspacefolder property to the CustomBuildTaskTerminal class.

This is necessary to make each task run in the correct project

* Simplify getAllCMakeProjects method.

It is a simple delegation to the eponym method of projectController

* The scope is now always a workspaceFolder

* Adds the getDefaultPresetNameForProject function

* Renames getDefaultPresetName function into getDefaultPresetNameForActiveProject

* Renames getActiveProject method into getProject method

Using workspaceFolder scope allows to retrieve the project using this folder

* Adds doc

* Get the preset for the specified workspace using the new getDefaultPresetNameForProject function

* Fix trailing space

* Adds comment

* Adds the method getProjectForFolder in the extension.

It is mostly a delegation to the eponym method of the projectController

* Update CHANGELOG.md (#3232)

---------

Co-authored-by: Garrett Campbell <86264750+gcampbell-msft@users.noreply.github.com>

Nerixyz committed:
feat: add `cmake.ctest.debugLaunchTarget` (#4273)

* feat: add `cmake.ctest.testSuiteDelimiter.debugLaunchTarget`

* fix: setting name

* fix: description name

* chore: add changelog entry

---------

Co-authored-by: Garrett Campbell <86264750+gcampbell-msft@users.noreply.github.com>

rjaegers committed:
fix(ctest): don't require backtrace in presence of DEF_SOURCE_LINE (#4321)

* ctest: Don't require backtrace in presence of DEF_SOURCE_LINE

* docs: add CHANGELOG entry

---------

Co-authored-by: Garrett Campbell <86264750+gcampbell-msft@users.noreply.github.com>

tringenbach committed:
Fix lowercase include paths on Mac (#4325)

* Fix lowercase include paths on Mac

* Update CHANGELOG for version 1.21

---------

Co-authored-by: Garrett Campbell <86264750+gcampbell-msft@users.noreply.github.com>

0xemgy committed:
Fix gnuld error parsing false positive on make errors, false negative due to trailing \r, and false parsing of new "multiple definitions" error (#4341)

* Fix gnuld error parsing false positive on make errors, false negative due to trailing \r, and false parsing of new "multiple definitions" error

* Move bug fix entry in CHANGELOG to 1.21 section

snehara99 committed:
Dev/snehara/update readme (#4338)

* graphics

* more images

* readme updates

* whitespace

* html sanitization

* made picture clarity better

* Update README.md

* Update README.md

* rewording suggestion from Garrett

Co-authored-by: Garrett Campbell <86264750+gcampbell-msft@users.noreply.github.com>

* rewording suggestion from Garrett

Co-authored-by: Garrett Campbell <86264750+gcampbell-msft@users.noreply.github.com>

* rewording suggestion from Garrett

Co-authored-by: Garrett Campbell <86264750+gcampbell-msft@users.noreply.github.com>

* rewording suggestion from Garrett

Co-authored-by: Garrett Campbell <86264750+gcampbell-msft@users.noreply.github.com>

* reordered features

* rewording and fixing broken contributing link

* removed extra period

* made feature and banner updates

* updated alt text

* reverted wording

* removed cache editor feature image

* updated wording for presets support feature

* removed extra image

---------

Co-authored-by: Garrett Campbell <86264750+gcampbell-msft@users.noreply.github.com>

gcampbell-msft committed:
fix getProjectForFolder (#4348)

TSonono committed:
Don't require LCOV to be found in test project (#4351)

snehara99 committed:
Update README.md (#4350)

* Update README.md

added a brief description of the features we support

* updated readme

gcampbell-msft committed:
be more explicit in what we check (#4352)

gcampbell-msft committed:
Localization - Translated Strings (#4344)

Co-authored-by: csigs <csigs@users.noreply.github.com>

hippo91 committed:
Enable test explorer for those test suites

hippo91 committed:
Keeping integrity of the return code as suggested in the review comments

hippo91 commented on src/cmakeProject.ts:
```diff
@@ -2331,7 +2331,7 @@ export class CMakeProject {
         const drv = await this.preTest(fromWorkflow);
         const retc = await this.cTestController.runCTest(drv);
-        return (retc) ? 0 : -1;
+        return (retc === 0) ? 0 : -1;
```
@gcampbell-msft i did change to keep integrity of the return code in the `ctest` method.

I am not confident to do the same change in the `cpack` method because i do not know how to check the result.

hippo91 commented on test/end-to-end-tests/single-root-ctest/project-folder/CMakePresets.json:
```diff
@@ -0,0 +1,44 @@
+{
+  "version": 2,
```
:+1:

hippo91 committed:
Takes into account code review suggestions

hippo91 commented on test/end-to-end-tests/single-root-ctest/project-folder/test_utils.cpp:
```diff
@@ -0,0 +1,31 @@
+ * @return int : 0 if the test is successful, 1 otherwise
+ */
+/********************************************************************************/
+int generic_test(const std::string& test_filepath, const bool success) {
```
:+1:

hippo91 committed:
Specify compilers for linux and windows systems.

hippo91 committed:
Select the preset according to the OS type

hippo91 commented:
@gcampbell-msft i integrated these tests in the PR build pipelines.

I configured C and CXX compilers for linux and windows systems. I do not have any windows os at home so i'm not able to test that. Neither can i on Mac OS. I was counting on the CI to check that.

gcampbell-msft committed:
Localization - Translated Strings (#4353)

Co-authored-by: csigs <csigs@users.noreply.github.com>

hippo91 commented:
@gcampbell-msft i tried to debug the failing CI tests but was unable to understand what was failing in the build step.

I suppose this is the build of the C++ test project but cannot confirm.
For now i try to run the CI locally using `act`. If you have any idea or hint about how i can debug the CI i'll be happy to try them.

hippo91 committed:
Merge branch 'main' into add_end_to_end_ctest_tests

hippo91 commented:
@gcampbell-msft i am trying to investigate but i do not know how i can have access to build output or log. Have you got any idea?
Another approach that i'am trying is to run the CI locally thanks to `act`. Maybe is there another possibility?

gcampbell-msft commented:
@hippo91 Are you not able to view the build log if you click on the github links? 

Our CI pipelines run publicly in github. Here is a link to the Windows one, for example, https://github.com/microsoft/vscode-cmake-tools/actions/runs/14145701868/job/40095177801?pr=4186

hippo91 commented:
@gcampbell-msft i am able to see the build log but for linux the only interesting thing is ```Warning: ls] 2025-04-07T11:17:46.863Z [warning] [proc] The command: /home/runner/work/_temp/-1604695791/cmake-3.18.3-Linux-x86_64/bin/cmake --build /home/runner/work/vscode-cmake-tools/vscode-cmake-tools/test/end-to-end-tests/single-root-ctest/project-folder/build -- exited with code: 1``` but i have no access to the inner log of this command.

It seems like the `cmake` command does not receive the correct config preset.
It is really weird because on my laptop (running debian) using `act`, the CI is successful.

gcampbell-msft commented:
@hippo91 I'm also interested in the failure for the Windows build. However, I also can't see any more information. I'll do my best to make time to investigate this sometime in the next week or so

hippo91 commented:
@gcampbell-msft the problem with windows is due to the fact i dumbly hardcoded path to /tmp which of course does not exist on windows. I should be able to fix this. However the only way for me to check windows execution is through the CI.

And for Linux in the CI for now i have no clue.

hippo91 committed:
Merge branch 'main' into add_end_to_end_ctest_tests

hippo91 committed:
Changes the location of temp directory holding the tests (TESTS_DIR).

Using /tmp is not compatible with windows use

hippo91 commented:
@gcampbell-msft i think i fixed the windows problem. Can you relaunch the CI please?

hippo91 committed:
Remove useless CMAKE variable

hippo91 commented:
@gcampbell-msft surprisingly linux ci seems to be ok.

I removed an unused CMAKE variable for windows. It was generating what could be an error.

hippo91 commented:
@gcampbell-msft the linux part of the CI is ok but i'm struggling with the windows one (the macos one is failing in an unrelated test). If you could have a look it would be nice. Please tell me if i can do anything to help.

gcampbell-msft committed:
Merge branch 'main' into add_end_to_end_ctest_tests

gcampbell-msft commented:
@hippo91 I updated to main and re-ran, it looks like the build for macos is failing even before running the tests, unfortunately I won't be able to investigate today, but we still hope to take this PR and we really appreciate your hard work!

Commit:
Merge remote-tracking branch 'origin/main' into add_end_to_end_ctest_tests

# Conflicts:
#	src/cmakeProject.ts

Commit:
Adapt single-root-ctest end-to-end tests to current API and fix Windows portability

 * Adopt CommandResult.exitCode API for cmake.ctest assertions (cmakeProjProject.ctest() now returns Promise<CommandResult>)
 * Set CMT_TESTING_OVERRIDE_TEST_EXPLORER=1 in launch.json so refreshTests in test mode falls through to runCTestImpl
 * Resolve cleanUpTestResultFiles ${workspaceFolder} macro at runtime instead of letting it land literally in y in rimraf
 * Pin Windows generator to "Visual Studio 17 2022" (windows-2022 runner no longer ships VS 16 2019)
 * Set CXX_STANDARD 17 on generated test executables (MSVC default is C++14 and the sources use <filesystem>)
 * Use path::string() in generate_output_file.cpp (path::string_type is std::wstring on Windows)
 * Set WINDOWS_EXPORT_ALL_SYMBOLS on TestUtils so the .lib import library is generated on MSVC

Commit:
enhance end-to-end CTest tests: update paths for cross-platform compatibility, improve directory handling, and refactor test configuration management.

Commit:
Merge branch 'main' into add_end_to_end_ctest_tests

Commit:
Merge branch 'main' into add_end_to_end_ctest_tests

Commit:
cleanUpTestResultFiles now derives the test result directory from the configure preset's `cacheVariables.TESTS_DIR` (expanding ${sourceDir})
   rather than a hardcoded literal that could silently drift, and guards
   the rmdir with `fs.exists` so the very first setup() — when the
   directory is absent — does not throw.
 * Snapshot the workspace-folder values of `cmake.ctest.allowParallelJobs`
   and `cmake.ctest.testSuiteDelimiter` once before any suite runs and
   restore them once after every suite has finished, so the project's
   checked-in `.vscode/settings.json` is no longer left dirty by a test
   run. Restoration intentionally happens in a global `after` hook (not
   per-suite) because writing the setting fires the extension's
   `refreshTests` cascade; deferring it to the very end avoids the
   cascade racing with the next suite's `BuildDirectoryHelper.clear()`
   on Windows (EBUSY) and isolates the shutdown-time build attempt to a
   scoped try/catch.

Commit:
lint errors

hippo91 commented:
@hanniavalera @gcampbell-msft thanks for your work. I am happy to see this finally merged!
